crates/cadenza-gcode/src/gcode_syntax.rs

```diff
@@ -0,0 +1,225 @@
+
+/// Parse GCode source into a Cadenza-compatible AST.
+pub fn gcode_parse(src: &str) -> Parse {
+    GCodeParser::new(src).parse()
+}
+
+struct GCodeParser<'src> {
+    src: &'src str,
+    offset: usize,
+    builder: GreenNodeBuilder<'static>,
+    errors: Vec<cadenza_syntax::parse::ParseError>,
+}
+
+impl<'src> GCodeParser<'src> {
+    fn new(src: &'src str) -> Self {
+        Self {
+            src,
+            offset: 0,
+            builder: GreenNodeBuilder::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    fn parse(mut self) -> Parse {
+        self.builder.start_node(Kind::Root.into());
+
+        for line in self.src.lines() {
+            let line_start = self.offset;
+            self.parse_line(line, line_start);
+            self.offset = line_start + line.len() + 1; // +1 for newline
+        }
+
+        self.builder.finish_node();
+
+        Parse {
+            green: self.builder.finish(),
+            errors: self.errors,
+        }
+    }
+
+    fn parse_line(&mut self, line: &str, line_start: usize) {
+        let line = line.trim();
+
+        if line.is_empty() {
+            return;
+        }
+
+        // Handle comments
+        if line.starts_with(';') {
+            // Comments are trivia, we can skip them or add as whitespace
+            return;
+        }
+
+        // Handle inline comments
+        let code_part = if let Some(pos) = line.find(';') {
+            line[..pos].trim()
+        } else {
+            line
+        };
+
+        if code_part.is_empty() {
+            return;
+        }
+
+        // Parse command as an Apply node
+        self.parse_command(code_part, line_start);
+    }
+
+    fn parse_command(&mut self, code: &str, line_start: usize) {
+        let parts: Vec<&str> = code.split_whitespace().collect();
+        if parts.is_empty() {
+            return;
+        }
+
+        // Start an Apply node for the command
+        self.builder.start_node(Kind::Apply.into());
+
+        // Parse command code (G1, M104, etc.) as the receiver
+        self.builder.start_node(Kind::ApplyReceiver.into());
+        self.parse_command_identifier(parts[0], line_start);
+        self.builder.finish_node();
+
+        // Parse parameters as arguments
+        for param in &parts[1..] {
+            self.builder.start_node(Kind::ApplyArgument.into());
+            self.parse_parameter(param, line_start);
+            self.builder.finish_node();
+        }
+
+        self.builder.finish_node();
+    }
+
+    fn parse_command_identifier(&mut self, cmd: &str, _line_start: usize) {
+        // Create an Identifier node (not just a token)
+        self.builder.start_node(Kind::Identifier.into());
+        self.builder.token(Kind::Identifier.into(), cmd);
+        self.builder.finish_node();
+    }
+
+    fn parse_parameter(&mut self, param: &str, _line_start: usize) {
+        if param.is_empty() {
+            return;
+        }
+
+        let letter = param.chars().next().unwrap();
```

Comment:
It would be great if the AST produced `[=, key, value]` pairs for arguments. That would make the arguments non-positional and give more information to the handler macro about what parameter it's expected to be.